File: server/src/main/java/com/krillsson/sysapi/core/metrics/windows/WindowsGpuMetrics.java
package com.krillsson.sysapi.core.metrics.windows;

import com.krillsson.sysapi.core.domain.gpu.Gpu;
import com.krillsson.sysapi.core.domain.gpu.GpuHealth;
import com.krillsson.sysapi.core.domain.gpu.GpuLoad;
import com.krillsson.sysapi.core.metrics.defaultimpl.DefaultGpuMetrics;
import com.krillsson.sysapi.util.Streams;
import oshi.hardware.HardwareAbstractionLayer;

import java.util.List;
import java.util.stream.Collectors;

import static com.krillsson.sysapi.core.metrics.windows.util.NullSafeOhmMonitor.nullSafeGetValue;

public class WindowsGpuMetrics extends DefaultGpuMetrics {

    private final DelegatingMonitorManager monitorManager;


    public WindowsGpuMetrics(HardwareAbstractionLayer hal, DelegatingMonitorManager monitorManager) {
        super(hal);
        this.monitorManager = monitorManager;
    }

    @Override
    public List<Gpu> gpus() {
        monitorManager.update();
        return Streams.ofNullable(monitorManager.GpuMonitors()).map(m -> new Gpu(
                m.getVendor(),
                m.getName(),
                nullSafeGetValue(m.getCoreClock()),
                nullSafeGetValue(m.getMemoryClock())
        )).collect(Collectors.toList());
    }

    @Override
    public List<GpuLoad> gpuLoads() {
        monitorManager.update();
        return Streams.ofNullable(monitorManager.GpuMonitors())
                .map(g -> new GpuLoad(
                        g.getName(), nullSafeGetValue(g.getCoreLoad()),
                        nullSafeGetValue(g.getMemoryLoad()),
                        new GpuHealth(
                                nullSafeGetValue(g.getFanRPM()),
                                nullSafeGetValue(g.getFanPercent()),
                                nullSafeGetValue(g.getTemperature())
                        )
                )).collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/krillsson/sysapi/core/metrics/windows/WindowsGpuMetrics.kt
package com.krillsson.sysapi.core.metrics.windows

import com.krillsson.sysapi.core.domain.gpu.Gpu
import com.krillsson.sysapi.core.domain.gpu.GpuHealth
import com.krillsson.sysapi.core.domain.gpu.GpuLoad
import com.krillsson.sysapi.core.metrics.defaultimpl.DefaultGpuMetrics
import com.krillsson.sysapi.core.metrics.windows.util.NullSafeOhmMonitor
import org.springframework.stereotype.Component
import oshi.hardware.HardwareAbstractionLayer
import java.util.stream.Collectors

@Component
class WindowsGpuMetrics(hal: HardwareAbstractionLayer, private val monitorManager: OHMManager) :
    DefaultGpuMetrics(hal) {
    override fun gpus(): List<Gpu> {
        monitorManager.update()
        return monitorManager.gpuMonitors().map { monitor ->
            Gpu(
                monitor.getVendor(),
                monitor.getName(),
                NullSafeOhmMonitor.nullSafeGetValue(monitor.getCoreClock()),
                NullSafeOhmMonitor.nullSafeGetValue(monitor.getMemoryClock())
            )
        }
    }

    override fun gpuLoads(): List<GpuLoad> {
        monitorManager.update()
        return monitorManager.gpuMonitors().map { monitor ->
            GpuLoad(
                monitor.getName(), NullSafeOhmMonitor.nullSafeGetValue(monitor.getCoreLoad()),
                NullSafeOhmMonitor.nullSafeGetValue(monitor.getMemoryLoad()),
                GpuHealth(
                    NullSafeOhmMonitor.nullSafeGetValue(monitor.getFanRPM()),
                    NullSafeOhmMonitor.nullSafeGetValue(monitor.getFanPercent()),
                    NullSafeOhmMonitor.nullSafeGetValue(monitor.getTemperature())
                )
            )
        }

    }
}
